### DISCUSSION TITLE
Link has leaky abstraction

### DISCUSSION THREAD

**Discussion by wereHamster** [vercel/next.js#11769] [Help] [answered]
I have a website that consistent of individual pages (/page/1, /page/2, /page/3, …). I used links in the form of `<Link href="/page/2">` to link to other pages. Then Next.js 9.3 came out and I refactored the individual pages into a single page (/page/[id].js) with getStaticProps. And now none of the links work correctly, because it is not possible to link to pages which use path params with `<Link href="…">`, I'd have to use `<Link href={{ pathname: "…", query: { … } }}>`. 

The Link API as it currently is leaks the details of how the page is implemented. I, as a developer, who wants to link to another page, don't actually care how the other page is implemented–standalone page vs page which uses getStaticProps–I just want to link to `/page/foo`.

I see a couple options how to fix this:

1. You make available all the necessary information about existing pages so that `<Link>` can at runtime parse the href into pathname + query.
2. You make `<Link>` a babel macro which can at build time parse the href into pathname + query.

**Comment by mliq** [comment, marked as answer]
I believe the correct usage is:

`<Link href='/page/[id]' as='/page/2'>`

I am not totally clear if this solves your problem or exactly what the problem is but hope that might be helpful.

**Reply by Timer** [reply]
We have an RFC to remove the necessity of configuring both `href` and `as`:
#8207
